src/transformers/models/nucleus_x/modeling_nucleus_x.py

```diff
@@ -0,0 +1,1441 @@
+        recurrent_chunk_size = (
+            recurrent_chunk_size if recurrent_chunk_size is not None else self.config.recurrent_chunk_size
+        )
+
+        outputs = self.model(
+            input_ids,
+            attention_mask=attention_mask,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            forward_mode=forward_mode,
+            use_cache=use_cache,
+            recurrent_chunk_size=recurrent_chunk_size,
+            retention_rel_pos=retention_rel_pos,
+        )
+
+        hidden_states = outputs[0]
+        logits = self.lm_head(hidden_states)
+
+        loss = None
+        if labels is not None:
+            # Shift so that tokens < n predict n
+            shift_logits = logits[..., :-1, :].contiguous()
+            shift_labels = labels[..., 1:].contiguous()
+            # Flatten the tokens
+            loss_fct = nn.CrossEntropyLoss()
+            shift_logits = shift_logits.view(-1, self.config.vocab_size)
+            shift_labels = shift_labels.view(-1)
+            # Enable model parallelism
+            shift_labels = shift_labels.to(shift_logits.device)
+            loss = loss_fct(shift_logits, shift_labels)
+
+            if self.config.z_loss_coeff > 0:
+                # z_loss from PaLM paper
+                # z_loss = 1e-4 * log(log(z)), where z = sum(exp(logits))
+                z_loss = torch.logsumexp(shift_logits, dim=-1).log().mean()
+                loss += self.config.z_loss_coeff * z_loss
+
+        if not return_dict:
+            output = (logits,) + outputs[1:]
+            return (loss,) + output if loss is not None else output
+
+        return NucleusXCausalLMOutputWithPast(
+            loss=loss,
+            logits=logits,
+            past_key_values=outputs.past_key_values,
+            hidden_states=outputs.hidden_states,
+            retentions=outputs.retentions,
+            attentions=outputs.attentions,
+        )
+
+    def _crop_past_key_values(model, past_key_values, maximum_length):
+        """Since NucleusX's kv do not have length, no need to crop. Just return"""
+        return past_key_values
+
+    def prepare_inputs_for_generation(
+        self, input_ids, past_key_values=None, attention_mask=None, inputs_embeds=None, **kwargs
+    ):
+        # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
+        if inputs_embeds is not None and past_key_values is None:
+            model_inputs = {"inputs_embeds": inputs_embeds}
+        else:
+            model_inputs = {"input_ids": input_ids}
+
+        forward_mode = kwargs.get("forward_mode", "parallel")
+        if past_key_values is not None:
+            # NOTE: when we have past_key_values, using recurrent mode will be faster.
```

Comment:
@Rocketknight1 @gante For your question about "generate code handling networks with multiple forward mode" in another comment, this is my take on that: when we call `prepare_inputs_for_generation`, if we detect `past_key_values`, it means that the prompt to the generation has been computed, and using recurrent forward is better. Hence the `forward_mode = "recurrent"` line below.

Note that `forward_mode` is just a string that the model takes as forward input to select the forward mode at each forward step!

Reply:
@Rocketknight1 @syncdoth We can support multiple generation models, but the implementation for it depends on a few factors! The audio models are the best examples. For instance:
1. Whisper wraps `generate` and accepts additional flags. These flags trigger additional arguments for `generate` ([e.g. a custom logits processor to generate timestamps](https://github.com/huggingface/transformers/blob/b86c54d9ff4504b7287d95a215f8d6fa9388761f/src/transformers/models/whisper/modeling_whisper.py#L1965)) or postprocessing
2. Bark contains 3 internal models, and wraps `generate` to call `generate` on them in a sequence

Additionally, if `model.forward` accepts multiple modes, you can also prepare the flags in `model. prepare_inputs_for_generation`, as written above :)